brave/src/main/java/brave/handler/SpanCollector.java

```diff
@@ -0,0 +1,134 @@
+ */
+public interface SpanCollector {
+  /** What ended the data collection? */
+  enum Cause {
+    /**
+     * Called on {@link Span#abandon()}.
+     *
+     * <p>This is useful when counting children. Decrement your counter when this occurs as the
+     * span will not be reported.
+     *
+     * <p><em>Note:</em>Abandoned spans should be ignored as they aren't indicative of an error.
+     * Some instrumentation speculatively create a span for possible outcomes such as retry.
+     */
+    ABANDON,
+    /**
+     * Called on {@link Span#finish()} and is the simplest cause to reason with. When {@link
+     * MutableSpan#startTimestamp()} is present, you can assume with high confidence you have all
+     * recorded data for this span.
+     */
+    FINISH,
+    /**
+     * Called on {@link Span#flush()}.
+     *
+     * <p>Even though the span here is incomplete (missing {@link MutableSpan#finishTimestamp()},
+     * it is reported to the tracing system unless a {@link FinishedSpanHandler} returns false.
+     */
+    FLUSH,
+    /**
+     * Called when the trace context was garbage collected prior to completion.
+     *
+     * <p>Unlike {@link FinishedSpanHandler#supportsOrphans()}, this is called even if empty.
+     * Non-empty spans are reported to the tracing system unless a {@link FinishedSpanHandler}
+     * returns false.
+     *
+     * @see FinishedSpanHandler#supportsOrphans()
+     */
+    ORPHAN
+  }
+
+  /** Use to avoid comparing against null references */
+  SpanCollector NOOP = new SpanCollector() {
+    @Override public void begin(TraceContext context, MutableSpan span, TraceContext parent) {
+    }
+
+    @Override public boolean end(TraceContext context, MutableSpan span, Cause cause) {
+      return true;
+    }
+
+    @Override public String toString() {
+      return "NoopSpanCollector{}";
+    }
+  };
+
+  /**
+   * This is called when a span is sampled, but before it is started.
+   *
+   * @param context the trace context which is  {@link TraceContext#sampledLocal()}. This includes
+   * identifiers and potentially {@link TraceContext#extra() extra propagated data} such as baggage
+   * or extended sampling configuration.
+   * @param span a mutable object that stores data recorded with span apis. Modifications are
+   * visible to later collectors.
+   * @param parent can be {@code null} only when the new context is a {@linkplain
+   * TraceContext#isLocalRoot() local root}.
+   */
+  void begin(TraceContext context, MutableSpan span, @Nullable TraceContext parent);
+
+  /**
+   * Called when data collection completes for one of the following reasons:
+   *
+   * <p><ol>
+   * <li>{@link Cause#ABANDON} if it was a speculative context</li>
+   * <li>{@link Cause#FINISH} if it was reported complete</li>
+   * <li>{@link Cause#FLUSH} if it was intentionally reported incomplete</li>
+   * <li>{@link Cause#ORPHAN} if it was reported incomplete due to garbage collection</li>
```

Comment:
does this need to be duplicated from `Cause` ? 

`Called when data collection completes` on its own is enough, and maybe link in the `cause` parameter to the enum.

Reply:
sgtm I can put this on the type docs of Cause, as I do think an overview helps, but possibly not here..